gardners/surveysystem
# Add backend fields: unit

>please merge after PR #86 is resolved

see #72 
# unit field
* adds **unit** field (string).  Since there is no logic  in the backend to this field I refrained from eunums. it behaves similar to the `text` field

| field     | in question | in answer | description |
| ---         | ---         | ---       | ---         |
| unit      | x           | x         | defines, displays and stores value unit |

Formats:

* question: `question1:How many days are in February?:How many days are in February?:INT:0::-1:-1:0:0::days`, leave empty (::) if no unit is required
* next_question: `{"next_questions": [{"id": "question1", "name": "question1", "title": "How many days are in February?", "title_text": "How many days are in February?", "type": "INT", "unit": "days"}]}`
* answer: `question1::28:0:0:0:0:0:0:days`



see [field_unit.test](https://github.com/gardners/surveysystem/blob/421675cb15b3038535390c58dfcfb30930fb4044/backend/tests/field_unit.test)

Apologies, this PR is polluted by some frontend commits and some automatic tabs-for-spaces replacement by my editor

@gardners Still learning.. Here is the workflow I followed for adding a new field. Please check and give feedback if I forgot something so we can document this

# Workflow for adding a field

### question

1. **includes/survey.h**: 
    - add field to `struct question`
2. **src/serialisers.c** 
    - add field to `serialise_question()`
    - add field to `deserialise_question()`
    - add field to `compare_questions()`
3. **src/sessions.c**
    - `free_question()` add `freez(<field>)`
4. **src/fcgimain.c**
    - `fcgi_nextquestion()` add `kjson_put_*` declaration for field

### answer (only if field is stored in answer)

1. **includes/survey.h**: 
    - add field to `struct answer`
2. **src/serialisers.c** 
    - add field to `serialise_answer()`
    - add field to `deserialise_answer()`
    - add field to `compare_answers()`
**???** Note: in my understanding there is no need to add the field to `free_answer()`?


### debug

* **src/serialisers.c**
 int dump_question

### tests

1. add field to existing tests
    - payload
    - next_question
    - answer check

1. create field_<my field>.test test
    - payload
    - next_question
    - answer check
